### DISCUSSION TITLE
Wrapping/hiding nlohmann::json implementation

### DISCUSSION THREAD

**Discussion by damiano-lonigo** [nlohmann/json#3775] [Q&A] [answered]
Hello, 
I am trying to hide the chosen json implementation using a wrapper class.
Normally there is a forward declaration in the .h something like

``
namespace nlohmann { class json; }

class Pjson
	{
	nlohmann::json *implP;
	
public:
	Pjson();
	virtual ~Pjson();
	};
``

and then in the .cpp there is

`#include <Pjson.h>
#include <json.hpp>

using Json = nlohmann::json;

Pjson::Pjson()
	{
	implP = new Json();
	}
`

I have a weird error **reference to 'json' is ambiguous**  

is there any way out of it ?

Thank you

**Comment by falbrechtskirchinger** [comment, marked as answer]
```cpp
namespace nlohmann { class json; }
```
This is not a correct forward declaration. Please include `json_fwd.hpp` which forward declares the nlohmann JSON types correctly.

**Reply by damiano-lonigo** [reply]
ok, downloaded, also added abi_macros.hpp and now I get **json.hpp:19273:7: error: redeclared with 10 template parameters**

**Reply by falbrechtskirchinger** [reply]
You're probably mixing versions or amalgamated and non-amalgamated headers. Download both `json.hpp` and `json_fwd.hpp` from the list of release artifacts here: https://github.com/nlohmann/json/releases/tag/v3.11.2
(Note that `include.zip` doesn't contain `json_fwd.hpp`. This was an oversight and it will be included in the next release.)

If you want to use the headers from `develop` choose between the amalgamated headers in the folder `single_include` or the individual headers in `include` and don't mix and match.

**Reply by falbrechtskirchinger** [reply]
> ok, downloaded, also added abi_macros.hpp and now I get **json.hpp:19273:7: error: redeclared with 10 template parameters**

This specific error is due to the upcoming release adding a tenth template parameter.

**Comment by damiano-lonigo** [comment]
yes, right, correct

when using the correct files, then, it works

just wondering if we are making the whole "programming environment" a bit too fragile

anyway, yes, thank you

**Reply by falbrechtskirchinger** [reply]
Well, just upgrade these files together, don't forward declare yourself, and everything should be fine.

`nlohmann::json` is just an alias and if that's the only type you care about, you should treat the `nlohmann:basic_json<...>` template as a black box and don't worry about it. The only time this would be expected to possibly break is the next major version release.

**Comment by damiano-lonigo** [comment]
I want to make your answer as a "solution" but I have the "mark answer" button on what I post and not on you post" ....

weird

**Reply by falbrechtskirchinger** [reply]
:man_shrugging:

Maybe @nlohmann can mark the answer for you.
